### DISCUSSION TITLE
How mirth will handle different thread counts between sequential channels.

### DISCUSSION THREAD

**Discussion by ghost** [nextgenhealthcare/connect#5296] [Q&A]
Scenario - 
1) Parent channel :  10 max processing threads at source. 500 messages queued up.
2) Downstream channel : 5 max processing threads at source and stopped

How downstream channel will process the 500 messages on start of the channel ? 

1) All 500 messages will get queued up at downstream channel straightaway and then 5 threads will process from downstream queue in parallel. 
 or 
2) No queuing will happen on downstream, messages from parent channel queue will be processed based on the downstream channel thread availability (out of 5) .

And also is there any way to view  thread related information for each message processed ?

**Comment by jonbartels** [comment]
You need to show the destination threading and queueing settings also.

**Reply by ghost** [reply]
Thanks @jonbartels .

Please find the destination queue threads and buffer size below:

 **Parent channel** - queue the messages at destination level.

| Max processing threads | Destination(Send to Downstream channel) <br> Queue Threads | Destination(Send to Downstream channel) <br>Queue Buffer Size |  
|-----------|:-----------:|-----------:|  
| <center>10</center> | <center>10 threads</center> | <center>1000 messages</center>|  

 **Downstream channel** 

| Max processing threads | Destination(None) <br>  Queue Threads | Destination(None) <br> Queue Buffer Size |  
|-----------|:-----------:|-----------:|  
| <center>5</center> | <center>1 threads</center> | <center>1000 messages</center>|

**Reply by jonbartels** [reply]
Thinking through this ....

Parent channel will receive 10 messages and run the transformers in parallel then queue those messages. The queue will process another 10 messages in parallel and send those to the Downstream channel. The Downstream channel will then have 5 threads to process transformers and **NOT QUEUE** (am I reading that right? queues off on Downstream?) .

There is a quirk in how channel writers work. The channel writers (and `router.routeMessage`) actually run the source transformations of the target channel in the destination thread of the sending channel. This means that the destination thread is slaved to the source thread of the receiving channel, the most common goof is that you'll get queueing on a parent channel when the target channel has a slow source transformer or where the target channel is not using queueing.

So to try and answer your actual questions:

* I think you'll see queues on your parent channel in the destination.
* If you see SOURCE queues, stop and think! Source queues are awful to manage and you should change your workflow to get destination queues.
* The only way i know of to view thread related information is to take a Java thread dump from MC server and see whats running. If you search the MC forums+GitHub for "Murmur" you'll find a great writeup from Nick Rupley on the hash algorithm and thread assignment mechanism that MC uses. https://github.com/nextgenhealthcare/connect/issues/3669

**Reply by ghost** [reply]
Thanks @jonbartels. Thanks for sharing your thoughts and referencing Murmur. 

Yes, I can see queue on parent channel at destination level.

**Reply by jonbartels** [reply]
What behavior do you want? Where do you want the queue to be for your use case?

Also note that the "quirk" only applies to channel writers, if your real use case is HTTP or TCP or something else it'll behave differently.
